# DataCamp: Analyzing Police Activity with Pandas

In [1]:
import pandas as pd

In [8]:
ri = pd.read_csv('data/police_ri.csv')
ri.head(3)


C:\Users\rbouc\AppData\Local\Temp\ipykernel_16744\2254315748.py:1: DtypeWarning: Columns (6,17,30) have mixed types. Specify dtype option on import or set low_memory=False.
  ri = pd.read_csv('data/police_ri.csv')


raw_row_number        date      time zone subject_race subject_sex  \
0               1  2005-11-22  11:15:00   X3        white        male   
1               2  2005-10-01  12:20:00   X3        white        male   
2               3  2005-10-01  12:30:00   X3        white      female   

  department_id       type arrest_made citation_issued  ... reason_for_stop  \
0           200  vehicular       False            True  ...        Speeding   
1           200  vehicular       False            True  ...        Speeding   
2           200  vehicular       False            True  ...        Speeding   

  vehicle_make vehicle_model raw_BasisForStop raw_OperatorRace  \
0          NaN           NaN               SP                W   
1          NaN           NaN               SP                W   
2          NaN           NaN               SP                W   

  raw_OperatorSex raw_ResultOfStop raw_SearchResultOne  raw_SearchResultTwo  \
0               M                M                 NaN                  NaN   
1               M                M                 NaN                  NaN   
2               F                M                 NaN                  NaN   

  raw_SearchResultThree  
0                   NaN  
1                   NaN  
2                   NaN  

[3 rows x 31 columns]

In [9]:
# look at the shape of the data
ri.shape

(509681, 31)

We see that there are 31 columns with 509,681 rows. 

Next, let's check for the presence of null values and clean up the data

In [3]:
# locate the missing values (1) - visual inspection
ri.isnull()

raw_row_number   date   time   zone  subject_race  subject_sex  \
0                False  False  False  False         False        False   
1                False  False  False  False         False        False   
2                False  False  False  False         False        False   
3                False  False  False  False         False        False   
4                False  False  False  False         False        False   
...                ...    ...    ...    ...           ...          ...   
509676           False  False  False  False         False        False   
509677           False  False  False  False         False        False   
509678           False  False  False  False         False        False   
509679           False  False  False  False         False        False   
509680           False   True   True   True          True         True   

        department_id   type  arrest_made  citation_issued  ...  \
0               False  False        False            False  ...   
1               False  False        False            False  ...   
2               False  False        False            False  ...   
3               False  False        False            False  ...   
4               False  False        False            False  ...   
...               ...    ...          ...              ...  ...   
509676          False  False        False            False  ...   
509677          False  False        False            False  ...   
509678          False  False        False            False  ...   
509679          False  False        False            False  ...   
509680           True  False         True             True  ...   

        reason_for_stop  vehicle_make  vehicle_model  raw_BasisForStop  \
0                 False          True           True             False   
1                 False          True           True             False   
2                 False          True           True             False   
3                 False          True           True             False   
4                 False          True           True             False   
...                 ...           ...            ...               ...   
509676            False         False          False             False   
509677            False          True           True             False   
509678            False         False          False             False   
509679            False         False          False             False   
509680             True          True           True              True   

        raw_OperatorRace  raw_OperatorSex  raw_ResultOfStop  \
0                  False            False             False   
1                  False            False             False   
2                  False            False             False   
3                  False            False             False   
4                  False            False             False   
...                  ...              ...               ...   
509676             False            False             False   
509677             False            False             False   
509678             False            False             False   
509679             False            False             False   
509680              True             True              True   

        raw_SearchResultOne  raw_SearchResultTwo  raw_SearchResultThree  
0                      True                 True                   True  
1                      True                 True                   True  
2                      True                 True                   True  
3                      True                 True                   True  
4                      True                 True                   True  
...                     ...                  ...                    ...  
509676                 True                 True                   True  
509677                 True                 True                   True  
50

In [11]:
# locate missing values (2) - sum up the number of missing values per column. 
ri.isnull().sum().sort_values()

raw_row_number                0
type                          0
search_conducted              0
date                         10
time                         10
zone                         10
department_id                10
frisk_performed              10
raw_OperatorRace          29073
raw_BasisForStop          29073
reason_for_stop           29073
raw_OperatorSex           29073
raw_ResultOfStop          29073
warning_issued            29073
citation_issued           29073
arrest_made               29073
subject_race              29073
subject_sex               29097
outcome                   35841
vehicle_make             191564
vehicle_model            279593
raw_SearchResultOne      491919
contraband_other         491919
search_basis             491919
reason_for_search        491919
contraband_found         491919
contraband_drugs         493693
contraband_weapons       497886
contraband_alcohol       508464
raw_SearchResultTwo      508862
raw_SearchResultThree    509513
dtype: i

We see that there are columns were nearly all values are null.

Let's drop 'raw_SearchResultThree' completely and then check the new shape of the data.

In [13]:
ri.drop('raw_SearchResultThree', axis='columns',inplace=True)

ri.shape

(509681, 30)

Next, let's drop rows where some critical values, such as search time, are missing.

In [16]:
ri.dropna(subset=['date', 'time'], inplace=True)

ri.shape

(509671, 30)